From 5653d6ea6a22391095fe40d7e693e8680786e6cb Mon Sep 17 00:00:00 2001
From: Schuyler Patton <spatton@ti.com>
Date: Thu, 30 Jun 2011 13:39:25 -0500
Subject: [PATCH 1/3] omap3_evm: Added function calls to set volts, speed on OMAP36xx parts

* evm.c

*  Added setup calls to set voltage and speed

*  Added calls in misc_init_r to twl4030_power_mpu_init and set_mpu_clock
   functions. These set VDD1 to 1.35v and mpu clock to 1GHz respectively.
   There is a check performed to make sure that this is only called on
   omap36xx class parts.

* twl4030.c

* Added twl4030_power_mpu_init function to set VDD1 to 1.35 Volts

* This is done to allow omap36xx parts to run at 1GHz.

* clock.c

* Added set_mpu_clk function to allow u-boot to set the clock
  rate.

* sys_proto.h

* Added the function prototype set_mpu_clock

* twl4030.h

* Added twl4030_power_mpu_init function prototype

Upstream-Status: Pending

Signed-off-by: Schuyler Patton <spatton@ti.com>
---
 arch/arm/cpu/arm_cortexa8/omap3/clock.c     |   15 +++++++++++++++
 arch/arm/include/asm/arch-omap3/sys_proto.h |    2 ++
 board/ti/evm/evm.c                          |    5 +++++
 drivers/power/twl4030.c                     |   12 ++++++++++++
 include/twl4030.h                           |    3 +++
 5 files changed, 37 insertions(+), 0 deletions(-)

diff --git a/arch/arm/cpu/arm_cortexa8/omap3/clock.c b/arch/arm/cpu/arm_cortexa8/omap3/clock.c
index d797dba..f21ab7b 100644
--- a/arch/arm/cpu/arm_cortexa8/omap3/clock.c
+++ b/arch/arm/cpu/arm_cortexa8/omap3/clock.c
@@ -35,6 +35,21 @@
 #include <command.h>
 
 /******************************************************************************
+ * Setting the Clock speed 
+ *
+ *****************************************************************************/
+void set_mpu_clk(int speed)
+{
+	struct prcm *prcm_base = (struct prcm *)PRCM_BASE;
+
+
+	/* Set M */
+	sr32(&prcm_base->clksel1_pll_mpu, 8, 11, speed);
+	/* lock mode */
+	sr32(&prcm_base->clken_pll_mpu, 0, 3, PLL_LOCK);
+}
+
+/******************************************************************************
  * get_sys_clk_speed() - determine reference oscillator speed
  *                       based on known 32kHz clock and gptimer.
  *****************************************************************************/
diff --git a/arch/arm/include/asm/arch-omap3/sys_proto.h b/arch/arm/include/asm/arch-omap3/sys_proto.h
index 605c974..70127ec 100644
--- a/arch/arm/include/asm/arch-omap3/sys_proto.h
+++ b/arch/arm/include/asm/arch-omap3/sys_proto.h
@@ -30,6 +30,8 @@ typedef struct {
 void prcm_init(void);
 void per_clocks_enable(void);
 
+void set_mpu_clk(int speed);
+
 void memif_init(void);
 void sdrc_init(void);
 void do_sdrc_init(u32, u32);
diff --git a/board/ti/evm/evm.c b/board/ti/evm/evm.c
index bffcbd6..635a68e 100644
--- a/board/ti/evm/evm.c
+++ b/board/ti/evm/evm.c
@@ -36,6 +36,7 @@
 #include <i2c.h>
 #include <asm/mach-types.h>
 #include "evm.h"
+#include <twl4030.h>
 
 static u8 omap3_evm_version;
 
@@ -98,6 +99,10 @@ int misc_init_r(void)
 
 	dieid_num_r();
 
+        if (is_cpu_family(CPU_OMAP36XX)) {
+   		twl4030_power_mpu_init();
+	        set_mpu_clk(500);
+	}
 	return 0;
 }
 
diff --git a/drivers/power/twl4030.c b/drivers/power/twl4030.c
index f354834..05be1a5 100644
--- a/drivers/power/twl4030.c
+++ b/drivers/power/twl4030.c
@@ -88,6 +88,18 @@ void twl4030_power_init(void)
 			     TWL4030_PM_RECEIVER_VDAC_DEDICATED);
 }
 
+#define VDD1_VSEL_135	0x3C
+
+void twl4030_power_mpu_init(void)
+{
+        unsigned char byte;
+
+        /* 1.4 Volts */
+        byte = VDD1_VSEL_135;
+        twl4030_i2c_write_u8(TWL4030_CHIP_PM_RECEIVER, byte,
+                             TWL4030_PM_RECEIVER_VDD1_VSEL);
+}
+
 void twl4030_power_mmc_init(void)
 {
 	unsigned char byte;
diff --git a/include/twl4030.h b/include/twl4030.h
index ec825fd..6cffb32 100644
--- a/include/twl4030.h
+++ b/include/twl4030.h
@@ -513,6 +513,9 @@ static inline int twl4030_i2c_read_u8(u8 chip_no, u8 *val, u8 reg)
  * Power
  */
 
+/*For setting VDD1  */
+void twl4030_power_mpu_init(void);
+
 /* For hardware resetting */
 void twl4030_power_reset_init(void);
 /* For initializing power device */
-- 
1.7.0.4

